File: q1/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n;    // length of string
        string s; // string
        cin >> n >> s;

        map<char, char> next; // map to mark current next element of each character
        map<char, char> prev; // map to mark current previous element of each character
        prev[s[0]] = '$';     // no previous element of 1st element $=trash value
        next[s[n - 1]] = '$'; // no next element of last element
        prev['$'] = s[n - 1];
        next['$'] = s[0];
        // maps are used like linked list here.. to make deletion and merging easy

        int ans = 0;
        // marking next and previous character to every character
        for (int i = 0; i < n - 1; i++)
        {
            next[s[i]] = s[i + 1];
            prev[s[i + 1]] = s[i];
        }

        int i = 97; // starting from character 'a'.

        while (i < 97 + n)
        {
            int u = i;
            while (next[char(u)] == char(u + 1))
            {
                u++;
            } // select substring with consecutive sorted characters

            // merge non selected substring .. as selected will be appended at the end of string p
            next[prev[char(i)]] = next[char(u)];
            prev[next[char(u)]] = prev[char(i)];

            i = u + 1; // next iteration will start from next character
            ans++;
        }

        cout << ans << endl;
    }
}

File: q1/solution.java
import java.util.*;

public class solution {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();

        while (t-- > 0) {
            int n = scanner.nextInt(); // length of string
            String s = scanner.next(); // string

            Map<Character, Character> next = new HashMap<>(); // map to mark current next element of each character
            Map<Character, Character> prev = new HashMap<>(); // map to mark current previous element of each character
            prev.put(s.charAt(0), '$'); // no previous element of 1st element $=trash value
            next.put(s.charAt(n - 1), '$'); // no next element of last element
            prev.put('$', s.charAt(n - 1));
            next.put('$', s.charAt(0));
            // maps are used like linked list here.. to make deletion and merging easy

            int ans = 0;
            // marking next and previous character to every character
            for (int i = 0; i < n - 1; i++) {
                next.put(s.charAt(i), s.charAt(i + 1));
                prev.put(s.charAt(i + 1), s.charAt(i));
            }

            int i = 97; // starting from character 'a'

            while (i < 97 + n) {
                int u = i;
                while (next.get((char) u) != null && next.get((char) u) == (char) (u + 1)) {
                    u++;
                } // select substring with consecutive sorted characters

                // merge non-selected substring .. as selected will be appended at the end of string p
                next.put(prev.get((char) i), next.get((char) u));
                prev.put(next.get((char) u), prev.get((char) i));

                i = u + 1; // next iteration will start from next character
                ans++;
            }

            System.out.println(ans);
        }

        scanner.close();
    }
}
